lib/puppet/provider/user/useradd.rb

```diff
@@ -89,8 +93,39 @@ def shell=(value)
     value.is_a? Integer
   end
 
-  verify :groups, "Groups must be comma-separated" do |value|
-    value !~ /\s/
+  verify :groups, "Groups must not contain spaces" do |value|
+    prop = @resource.parameters[:groups]
+    changes = prop.should_but_isnt + prop.is_but_shouldnt
+    changes.all? do |change|
+      change !~ /\s/
+    end
+  end
+
+  def groups=(value)
+    self.class.validate(:groups, value, self)
```

Comment:
The `verify` check is yielded a `value`, but the thing we verify is a bit different (the list of groups split on ','). Also if the check fails, we say "Invalid value group,another group: Groups must not contain spaces" instead of the individual group name that had the issue.

I think it would be cleaner to check for whitespace when the `groups=` method is called. Since the method is already collecting the set of current and desired values, and enumerating them, we could issue an error message for the specific group:

```
def groups=(value)
  prop = @resource.parameters[:groups]
  ...
  changes.each do |group|
    self.fail("Invalid value '#{group[:name]': Group must not contain spaces") if group[:name] =~ /\s/
    md = [ command(:gpasswd),
      group[:name],
      group[:action],
     @resource[:name] ]
    ...
  end
```